src/server/generic_family.cc

```diff
@@ -243,161 +251,182 @@ OpResult<RestoreArgs> RestoreArgs::TryFrom(const CmdArgList& args) {
+  serialized_value_.sticky = it->first.IsSticky();
+
+  return OpStatus::OK;
+}
+
+OpStatus Renamer::DelSrc(Transaction* t, EngineShard* shard) {
+  auto res = shard->db_slice().FindMutable(t->GetDbContext(), src_key_);
+  auto& it = res.it;
+
+  CHECK(IsValid(it));
+
+  res.post_updater.Run();
+  CHECK(shard->db_slice().Del(t->GetDbIndex(), it));
+  if (shard->journal()) {
+    RecordJournal(t->GetOpArgs(shard), "DEL", ArgSlice{src_key_}, 2);
   }
 
   return OpStatus::OK;
 }
 
-OpStatus Renamer::UpdateDest(Transaction* t, EngineShard* es) {
-  if (es->shard_id() != src_sid_) {
-    auto& db_slice = es->db_slice();
-    string_view dest_key = dest_res_.key;
-    auto res = db_slice.FindMutable(t->GetDbContext(), dest_key);
-    auto& dest_it = res.it;
-    bool is_prior_list = false;
-
-    if (IsValid(dest_it)) {
-      bool has_expire = dest_it->second.HasExpire();
-      is_prior_list = dest_it->second.ObjType() == OBJ_LIST;
-
-      if (src_res_.ref_val.ObjType() == OBJ_STRING) {
-        dest_it->second.SetString(str_val_);
-      } else {
-        dest_it->second = std::move(pv_);
-      }
-      dest_it->second.SetExpire(has_expire);  // preserve expire flag.
-      db_slice.UpdateExpire(t->GetDbIndex(), dest_it, src_res_.expire_ts);
-    } else {
-      if (src_res_.ref_val.ObjType() == OBJ_STRING) {
-        pv_.SetString(str_val_);
-      }
-      auto op_res =
-          db_slice.AddNew(t->GetDbContext(), dest_key, std::move(pv_), src_res_.expire_ts);
-      RETURN_ON_BAD_STATUS(op_res);
-      res = std::move(*op_res);
-    }
+OpStatus Renamer::DeserializeDest(Transaction* t, EngineShard* shard) {
+  auto& db_slice = shard->db_slice();
+  auto res = db_slice.FindMutable(t->GetDbContext(), dest_key_);
 
-    dest_it->first.SetSticky(src_res_.sticky);
+  auto& dest_it = res.it;
+  const bool is_prior_list = IsValid(dest_it) && dest_it->second.ObjType() == OBJ_LIST;
 
-    if (!is_prior_list && dest_it->second.ObjType() == OBJ_LIST && es->blocking_controller()) {
-      es->blocking_controller()->AwakeWatched(t->GetDbIndex(), dest_key);
-    }
-    if (es->journal()) {
-      OpArgs op_args = t->GetOpArgs(es);
-      string scratch;
-      // todo insert under multi exec
-      RecordJournal(op_args, "SET"sv, ArgSlice{dest_key, dest_it->second.GetSlice(&scratch)}, 2,
-                    true);
-      if (dest_it->first.IsSticky()) {
-        RecordJournal(op_args, "STICK"sv, ArgSlice{dest_key}, 2, true);
-      }
-      if (dest_it->second.HasExpire()) {
-        auto time = absl::StrCat(src_res_.expire_ts);
-        RecordJournal(op_args, "PEXPIREAT"sv, ArgSlice{dest_key, time}, 2, true);
-      }
-      RecordJournalFinish(op_args, 2);
+  int rdb_version = 0;
+  CHECK(VerifyFooter(serialized_value_.value, &rdb_version));
+
+  RestoreArgs restore_args;
+  restore_args.expiration_ = serialized_value_.expire_ts;
+  restore_args.abs_time_ = true;
+  restore_args.replace_ = true;
```

Comment:
Why not add a constructor to `RestoreArgs`? That way you don't need to hack yourself into the interface with the `friend` chore above

Reply:
Oh the `friend` was for `Renamer`. Nvm but the comment for the constructor still applies :)